### ISSUE TITLE
compute: SetMetadata does not work?

### ISSUE THREAD

**Issue by gyuho** [GoogleCloudPlatform/google-cloud-go#654]
I am trying to update metadata on existing instance, but this call doesn't work. Can anybody check?

```go
hcli := http.DefaultClient

items := make(map[string]string)
items["a"] = "test"
csrv, err := compute.New(hcli)

items := make([]*compute.MetadataItems, 0, len(items))
for k, v := range items {
    // make sure to copy as value before passing as reference!
    copied := v
    items = append(items, &compute.MetadataItems{Key: k, Value: &copied})
}
metadata := &compute.Metadata{Items: items}

_, err = compute.
    NewInstancesService(csrv).
    SetMetadata(projectID, zone, name, metadata).
    Context(ctx).
    Do()
```

Thanks.

**Comment by jba** [comment]
What is the value of `err`?

**Comment by gyuho** [comment]
@jba `err == nil`. I manually check the instance status in the GCE UI. And never gets updated.
Thanks.

**Comment by jba** [comment]
There are a couple of possibilities:

It's not clear from your snippet whether you're correctly creating the client. You don't check the error from `compute.New`, and you will get an error because you're not doing any auth. Try this instead:
```go
ctx := context.Background()
hc, _, err := transport.NewHTTPClient(ctx)
if err != nil {
    log.Fatal(err)
}
csrv, err := compute.New(hc)
if err != nil {
    log.Fatal(err)
}
```
Here, `transport` is the package `google.golang.org/api/transport`.

By the way, there is already an `InstancesService` in `csrv`, so you can write the call as
```go
op, err := csrv.Instances.SetMetadata(projectID, zone, instance, md).Context(ctx).Do()
if err != nil {
    log.Fatal(err)
}
```
The second potential problem is that the call is asynchronous. You must poll for the completion of `op`:
```go
for {
    op, err = csrv.ZoneOperations.Get(projectID, zone, op.Name).Context(ctx).Do()
    if err != nil {
        log.Fatal(err)
    }
    if op.Status == "DONE" {
        break
    }
    time.Sleep(1 * time.Second)                                                                                                                                                                                                                      
}
```

**Comment by gyuho** [comment]
@jba Thanks for feedback! I took out the error check from my code. My original snippet looked like this:

```go
package main

import (
	"context"
	"encoding/json"
	"fmt"
	"io/ioutil"
	"time"

	"golang.org/x/oauth2/google"
	compute "google.golang.org/api/compute/v1"
)

func main() {
	key, err := ioutil.ReadFile("/etc/gcp-key.json")
	if err != nil {
		panic(err)
	}
	credMap := make(map[string]string)
	if err := json.Unmarshal(key, &credMap); err != nil {
		panic(err)
	}
	projectID, ok := credMap["project_id"]
	if !ok {
		panic(fmt.Errorf("key has no project_id %q", string(key)))
	}
	jwt, err := google.JWTConfigFromJSON(key, compute.ComputeScope)
	if err != nil {
		panic(err)
	}

	ctx := context.Background()
	client := jwt.Client(ctx)

	csrv, err := compute.New(client)
	if err != nil {
		panic(err)
	}
	zone := "us-west1-a"
	instanceName := "gyuho1"

	tv := "value"
	metadata := &compute.Metadata{Items: []*compute.MetadataItems{{Key: "key", Value: &tv}}}
	op, err := csrv.Instances.
		SetMetadata(projectID, zone, instanceName, metadata).
		Context(ctx).
		Do()
	if err != nil {
		panic(err)
	}
	for {
		op, err = csrv.ZoneOperations.Get(projectID, zone, op.Name).Context(ctx).Do()
		if err != nil {
			panic(err)
		}
		if op.Status == "DONE" {
			break
		}
		fmt.Println("waiting...")
		time.Sleep(1 * time.Second)
	}
	fmt.Println("Done!")
}
```

I made some updates based on your feedback, but still doesn't work. Basically I read service account key from disk, and try to set metadata on existing instance. I do not get any errors from server. And all API calls succeed as well, but there is no change in the instance.

Could anybody reproduce it? I run this several times, but still no change. It could be just me?

Thanks!

**Comment by jba** [comment]
Reproduced. Thanks for the code.

The problem is that you must provide a fingerprint in the metadata, like so:
```go
   inst, err := csrv.Instances.Get(projectID, zone, instanceName).Context(ctx).Do()
    if err != nil {
        log.Fatal(err)
    }
    tv := "value"
    metadata := &compute.Metadata{
        Items:       []*compute.MetadataItems{{Key: "key", Value: &tv}},
        Fingerprint: inst.Metadata.Fingerprint,
    }
```
Without the fingerprint, or with an invalid fingerprint, it fails silently. The intent is to prevent data loss from updating metadata blindly, without taking into account what is currently there.

The fingerprint requirement is [documented](https://cloud.google.com/compute/docs/reference/latest/instances/setMetadata), but I would consider the silent failure to be a bug in the compute service. I filed https://issuetracker.google.com/issues/62541072.
